torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1454,6 +1454,252 @@ class AutogradLazyBackwardCompileInfo:
     saved_compile_context: Optional[CompileContext]
 
 
+def _raise_if_functorch_active():
+    # not ideal but prevent the user from seeing a nasty traceback - See #138422
+    stack = torch._C._functorch.peek_interpreter_stack()
+    torch._check(
+        stack is None,
+        lambda: (
+            "It looks like you're trying to call a compiled backward function within vmap/grad/vjp, "
+            "which isn't supported. Try wrapping vmap inside torch.compile, or skip compiling the "
+            "backward function."
+        ),
+    )
+
+
+# NOTE: this function must be torch._dynamo.allow_in_graph-able. Non tensor/symnode inputs must be constants.
+def _backward_prologue_functional(
+    ctx_saved_tensors, ctx_symints, metadata, maybe_subclass_metadata, *flat_args
+):
+    # Calling convention: we expect a grad_out passed to the backward:
+    # - for every output of the fw that does *not* alias an input or graph intermediate
+    # - for every updated_input generated by the fw that does *not* alias an input (aka only data-mutations)
+    # - for every graph intermediate that we need to use to generate an output later.
+    # The other outputs in the autograd.Function.forward that do *not* show up in the backward include:
+    # - outputs that alias inputs or graph intermediates
+    # - updated inputs due to metadata-only mutations.
+    # We need to return them in the forward, but ensure that they all do not get gradients in the backward,
+    # and we filter them out here before passing the remaining grad_outputs into the compiled backward.
+    _raise_if_functorch_active()
+
+    num_intermediate_bases = metadata.num_intermediate_bases
```

Comment:
One general question: how are you thinking the tradeoff between:

(1)  "make all of the metadata an input to the allow_in_graph region"

(2) "have the allow_in_graph region secretly close over any metadata that AOTAutograd needs"

I had thought that secretly-closing-over-any-metadata might be easier, since you don't need to worry about ensuring that dynamo can plumb the `CompiledFunction.metadata` and `CompiledFunction.maybe_subclass_metadata` into the allow_in_graph region. I would have thought that dynamo wouldn't know how to proxy those two objects correctly - but did you not run into any issues with it?

Reply:
Oh I am using that secretly closing trick you mentioned:
https://github.com/pytorch/pytorch/blob/8c684e9cfa11d566979dc5573de5bc9b6d4b9ca7/torch/_dynamo/compiled_autograd.py#L219-L229

Reply:
and yes, we do run into dynamo variable issues with some of the fields if we don't do this